### DISCUSSION TITLE
Does MLS support out-of-order message delivery decryption

### DISCUSSION THREAD

**Discussion by Tenari** [openmls/openmls#1427] [Q&A]
I understand how signal's [Double Ratchet](https://signal.org/docs/specifications/doubleratchet) can handle the out of order/delayed messages as detailed in "2.6. Out-of-order messages" but that's only for 1-1DMs

MLS is for groups, does it have the same out-of-order affordances?

**Comment by kkohbrok** [comment]
Good question. This is buried a bit, but MLS' [deletion schedule](https://www.rfc-editor.org/rfc/rfc9420.html#name-deletion-schedule) allows us to keep around secrets that have not yet been used for decryption. The [sender ratchets](https://www.rfc-editor.org/rfc/rfc9420.html#name-encryption-keys) that hang off of the MLS secret tree and that manage the secrets for message encryption/decryption is essentially the same as the [symmetric ratchet](https://www.signal.org/docs/specifications/doubleratchet/#symmetric-key-ratchet) used by the Signal protocol. This construction allows us to safely keep around the key material for messages that arrive out-of-order.

OpenMLS also supports out-of-order messages across epochs. You can control the number of epochs for which you want a group to keep around message encryption/decryption key material using the `max_past_epochs` field of the [`MlsGroupConfig`](https://docs.rs/openmls/latest/openmls/group/struct.MlsGroupConfig.html#).

EDIT: It's actually not that buried. Here: https://www.rfc-editor.org/rfc/rfc9420.html#name-delayed-and-reordered-appli

**Comment by kkohbrok** [comment]
Note that this only goes for application messages, but not for handshake messages, i.e. messages that deal with key material and group state management. For the latter, MLS requires the Delivery Service to provide a strict ordering as described in the [MLS architecture document](https://messaginglayersecurity.rocks/mls-architecture/draft-ietf-mls-architecture.html#name-delivery-of-messages).
